packages/block-editor/src/components/block-alignment-visualizer/index.js

```diff
@@ -0,0 +1,391 @@
+			}
+		}
+	}, [ dragPosition ] );
+
+	useEffect( () => {
+		const resolvedBlocklistElement =
+			blockListBlockElement ?? rootBlockListElement;
+		if ( ! focusedBlockElement || ! resolvedBlocklistElement ) {
+			return;
+		}
+
+		const { ownerDocument } = focusedBlockElement;
+		const { defaultView } = ownerDocument;
+
+		const update = () => {
+			setPopoverAnchor( {
+				ownerDocument,
+				getBoundingClientRect() {
+					const blockListRect =
+						resolvedBlocklistElement.getBoundingClientRect();
+					const focusedBlockRect =
+						focusedBlockElement.getBoundingClientRect();
+
+					// Produce a rect that has:
+					// - the horizontal positioning/height of the parent block.
+					// - the vertical positioning/height of the current block.
+					//
+					// These are the dimensions of our fake 'block list'.
+					return new defaultView.DOMRect(
+						blockListRect.x,
+						focusedBlockRect.y,
+						blockListRect.width,
+						focusedBlockRect.height
+					);
+				},
+			} );
+
+			setCoverElementStyle( {
+				position: 'absolute',
+				width: resolvedBlocklistElement.offsetWidth,
+				height: focusedBlockElement.offsetHeight,
+			} );
+		};
+
+		const resizeObserver = defaultView.ResizeObserver
+			? new defaultView.ResizeObserver( update )
+			: undefined;
+		resizeObserver?.observe( resolvedBlocklistElement );
+		resizeObserver?.observe( focusedBlockElement );
+		update();
+
+		return () => {
+			resizeObserver?.disconnect();
+		};
+	}, [ focusedBlockElement, blockListBlockElement, rootBlockListElement ] );
+
+	const focusedBlockAllowedAlignments = getValidAlignments(
+		getBlockSupport( focusedBlockName, 'align' ),
+		hasBlockSupport( focusedBlockName, 'alignWide', true )
+	);
+
+	// Allow override of `blockAllowedAlignments`. The image block doesn't use
+	// alignment block supports, so this allows the image block to directly
+	// declare what it supports.
+	const availableAlignments = useAvailableAlignments(
+		allowedAlignments ?? focusedBlockAllowedAlignments
+	);
+
+	const alignments = useMemo( () => {
+		return availableAlignments
+			.map( ( { name } ) => {
+				if ( name === 'none' ) {
+					return {
+						name,
+						label: __( 'Content width' ),
+					};
+				}
+				if ( name === 'wide' ) {
+					return {
+						name,
+						label: __( 'Wide width' ),
+						className: 'alignwide',
+					};
+				}
+				if ( name === 'full' ) {
+					return {
+						name,
+						label: __( 'Full width' ),
+						className: 'alignfull',
+					};
+				}
+				return null;
+			} )
+			.filter( ( alignment ) => alignment !== null );
+	}, [ availableAlignments, layout ] );
+
+	const contrastColor = useMemo( () => {
+		if ( ! focusedBlockElement ) {
+			return;
+		}
+
+		return focusedBlockElement.ownerDocument.defaultView
+			.getComputedStyle( focusedBlockElement )
+			.getPropertyValue( 'color' );
+	}, [ focusedBlockElement ] );
+
+	const popoverRef = useRef();
+
+	if ( availableAlignments?.length === 0 ) {
```

Comment:
I'd prefer not prematurely guarding against nullish values. So that we can get a faster feedback loop when it breaks. It's not being checked in the `alignments` memo so we will never reach this check anyway 😆 .